### DISCUSSION TITLE
Apostrophe 3.1.2

### DISCUSSION THREAD

**Discussion by abea** [apostrophecms/apostrophe#3236] [Release Notes]
This week's release is relatively small with mostly code quality work in Apostrophe 3. That is merely the tip of an iceberg that will be more completely revealed in the next release? Do icebergs get revealed? Did the metaphor break down almost immediately? No matter.

## Internationalization

Here we have the rest of our metaphorical iceberg. The team has been working hard on the first major feature set following the 3.0 release. Internationalization is hugely important for many Apostrophe users. Static text localization was part of A2 core, but dynamic content localization required the Workflow module for A2. As with Workflow's draft review, that too will be moved into Apostrophe 3 core.

Before saying more, here are rough definitions of the main concepts involved as we use them:

- **Internationalization** (i18n) encompasses the features, including the core API, that enable people to tailor Apostrophe applications for specific groups of people (most commonly through text translation).
- **Localization** (l10n) is the process of using those tools to customize the application. This includes translating text strings, but can also displaying different content on a page, converting currencies and number formats, and more.
- **Locales** in Apostrophe are the different versions of application content, including translated strings.
- **Static content** includes the user interface and things a developer might hard-code into templates. **Dynamic content** is what editors will add and update through the UI.

The W3C has a more complete description of these concepts that is worth a few minutes: [https://www.w3.org/International/questions/qa-i18n.en](https://www.w3.org/International/questions/qa-i18n.en). Understanding this is worthwhile if you are interested in digging into the code involved.

### Progress achieved

This week we made great progress adding the `@apostrophecms/i18n` module to A3 core with an API for static content localization. We've also begun moving core text strings into a default English locale file. If you want to start looking at the code for this, check out [the `static-i18n` code branch](https://github.com/apostrophecms/apostrophe/tree/static-i18n) in Github. There are also technical design issues for discussion of [static localization](https://github.com/apostrophecms/apostrophe/issues/3204) and [dynamic content localization](https://github.com/apostrophecms/apostrophe/issues/3205) were you can share your thoughts.

We also have designs well under way for the l10n interface for editors. This includes UI for switching locales as well as UI to clone pages and pieces from one locale to one or more others. Here are a couple of sneak peek (not final) images of that.

![Screenshots of the upcoming internationalization dropdown menu and locale document cloning interface](https://user-images.githubusercontent.com/1155984/125653422-c40c1651-14dc-4d86-97ea-27d67c9d24ee.png)

We're aiming to start releasing these features as a part of the next release in two weeks starting with static l10n. Dynamic content l10n will be next, with features rolled out over the next few releases.

It is worth noting that if you do not have multiple locales for your application (you only operate in one language and country) you will not have to do anything new. These features will be ready if you ever do need them.

And if you are interested in contributing translations for Apostrophe's user interface, please reach out to us in [the Github issue](https://github.com/apostrophecms/apostrophe/issues/3204), the Discord chat, or at hello@apostrophecms.com.

#### Changes
* Removes an unused method, `mapMongoIdToJqtreeId`, that was used in A2 but is no longer relevant.
* Removes deprecated and non-functional steps from the edit method in the `AposDocsManager.vue` component.
* Legacy migrations to update 3.0 alpha and 3.0 beta sites to 3.0 stable are still in place, with no functional changes, but have been relocated to separate source files for ease of maintenance. Note that this is not a migration path for 2.x databases. Tools for that are forthcoming.

#### Fixes
* Fix server crash due to oembed error when invalid url is passed.

#### Security Fixes
* Extract throttle logic from the POST login route to apply it to the `POST` login `totp` route.

#### [gatsby-source-apostrophe 1.0.1](https://www.npmjs.com/package/gatsby-source-apostrophe)

- Corrects the `apos-refresh` query parameter for page markup to `aposRefresh`.

#### [mechanic 1.8.0](https://www.npmjs.com/package/mechanic)

- Specify `ssl_prefer_server_ciphers` on, which results in more secure cipher choices being chosen first and an "A" rating from SLL Labs without micromanagement of cipher settings.
- Bug fix: ensure `site.backends` exists so a simple static site doesn't crash template generation. This bug was introduced in version 1.7.0.

**Comment by daslicht** [comment]
Does that also include something like blog posts or other written articles ?

**Reply by abea** [reply]
Localizing blog post content? The "dynamic content" localization will include that kind of thing. It's anything that a web content editor might add to the website. Work on that feature set will begin over the next few release cycles.

**Reply by daslicht** [reply]
Sounds great  ..."following"
